File: naoAluno/insercaoParcial.c
/**
 * Algoritmo de ordenacao por insercao
 * @author Max do Val Machado
 * @version 2 01/2015
 */
#include <stdio.h>
#include <math.h>
#include <time.h>
#include <stdlib.h>

#define n   10000
#define k   100

int array[n];
int i;
int j;

//Produz um arranjo ordenado de modo decrescente.
void decrescente() {
   for (i = 0; i < n; i++) {
      array[i] = n - 1 - i;
   }
}


//Produz um arranjo de numeros aleatorios.
void aleatorio() {
   for (i = 0; i < n; i++) {
      array[i] = rand() % 1000;
   }
}


//Mostrar os elemento de um arranjo.
void mostrar() {
   printf("[ ");

   for (i = 0; i < k; i++) {
      printf("%d ", array[i]);
   }

   printf("] \n");
}


// Algoritmo de ordenacao
void partialInsertionSort() {
   for (i = 1; i < n; i++) {
      int tmp = array[i];
      j = i - 1;

      if(i > k && array[k-1] > tmp){
         array[i] = array[k-1];
         j = k - 1;
      }
      while ((j >= 0) && (array[j] > tmp)) {
         array[j + 1] = array[j];
         j--;
      }
      array[j + 1] = tmp;
   }
}

int main() {
   srand(time(NULL));
   decrescente();
   //mostrar();		
   clock_t comeco = clock();
   partialInsertionSort();
   clock_t fim = clock();
   int total = (clock() - comeco) / CLOCKS_PER_SEC / 1000;

   mostrar();
   printf("Tempo para ordenar: %d ms.", total);
}

File: naoAluno/InsercaoParcial.java
/**
 * Algoritmo de ordenacao por insercao
 * @author Max do Val Machado
 * @version 2 01/2015
 */

class InsercaoParcial extends Geracao {

	/**
	 * Construtor.
	 */
   public InsercaoParcial(){
      super();
   }


	/**
	 * Construtor.
	 * @param int tamanho do array de numeros inteiros.
	 */
   public InsercaoParcial(int tamanho){
      super(tamanho);
   }


	/**
	 * Algoritmo de ordenacao parcial por insercao.
	 */
	public static void insercaoParcial(int k) {
		for (int i = 1; i < n; i++) {
			int tmp = array[i];
         int j = i - 1;

         if(i > k && array[k-1] > tmp){
            array[i] = array[k-1];
            j = k - 1;
         }
         while ((j >= 0) && (array[j] > tmp)) {
            array[j + 1] = array[j];
            j--;
         }
         array[j + 1] = tmp;
		}
	}


	public static void main(String[] args) {
      InsercaoParcial insercaoParcial = new InsercaoParcial(100);
      int k = 10;

		insercaoParcial.aleatorio();
		insercaoParcial.mostrar();
		
		long comeco = now();
		insercaoParcial.insercaoParcial(k);
		long fim = now();
		
		insercaoParcial.mostrar();
		insercaoParcial.mostrar(k);
		System.out.println("Tempo para ordenar: " + (fim-comeco)/1000.0 + " s.");
	}
}
